core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -774,6 +783,18 @@ private[spark] class MapOutputTrackerMaster(
     }
   }
 
+  def getAllMapOutputStatuses(shuffleId: Int): Array[MapStatus] = {
+    logDebug(s"Fetching all output statuses for shuffle $shuffleId")
+    shuffleStatuses.get(shuffleId) match {
+      case Some(shuffleStatus) =>
+        shuffleStatus.withMapStatuses { statuses =>
+          MapOutputTracker.checkMapStatuses(statuses, shuffleId)
+          statuses.clone
```

Comment:
I see your intention by calling this `clone` here but I do not think this is enough.
As the `MapStatus` is trait and not a case class in addition its implementations are mutable with a lot of `var` fields.

The `clone` on the `Array` is not a deep copy.

Reply:
It could not be enough if the metadata can mutate. But as I see we could solve all the problems with immutable metadata easily. So to be on the safe side please document we require the metadata to be immutable and introduce an `updateMetadata(meta: Option[Serializable])` method in `MapStatus`. Then we will be safe and all the use cases are covered.

(And you can use a case class for the Uber RSS's `MapTaskRssInfo`)